core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java

```diff
@@ -153,14 +158,24 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
                 if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
                     if (ig != null) {
                         if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
-                            if (!primaryIgSelected) {
-                                primaryIgSelected = true;
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
-                                instanceMetaDataEntry.setServer(Boolean.TRUE);
-                                LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                            if (terminatedPrimaryGwWhichShouldBeRestored.isPresent()) {
+                                if (terminatedPrimaryGwWhichShouldBeRestored.get().getDiscoveryFQDN().equals(instanceMetaDataEntry.getDiscoveryFQDN())) {
+                                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
+                                    instanceMetaDataEntry.setServer(Boolean.TRUE);
+                                } else {
+                                    LOGGER.info("Primary gateway is another instance not this one: {}", instanceMetaDataEntry.getInstanceId());
+                                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY);
+                                }
                             } else {
-                                LOGGER.info("Primary gateway was selected");
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY);
+                                if (!primaryIgSelected) {
+                                    primaryIgSelected = true;
+                                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
+                                    instanceMetaDataEntry.setServer(Boolean.TRUE);
+                                    LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                                } else {
+                                    LOGGER.info("Primary gateway was selected");
+                                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY);
+                                }
```

Comment:
could you please do a bit refactor and move at least these out to methods? this method is already large as hell

Reply:
please cover these with unit tests